packages/medusa/src/services/gift-card.ts

```diff
@@ -0,0 +1,280 @@
+        relations as (keyof GiftCard)[],
+        query
+      )
+
+      if (!giftCard) {
+        const selectorConstraints = Object.entries(selector)
+          .map((key, value) => `${key}: ${value}`)
+          .join(", ")
+
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Gift card with ${selectorConstraints} was not found`
+        )
+      }
+
+      return giftCard
+    })
+  }
+
+  /**
+   * Gets a gift card by id.
+   * @param giftCardId - id of gift card to retrieve
+   * @param config - optional values to include with gift card query
+   * @return the gift card
+   */
+  async retrieve(
+    giftCardId: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async () => {
+      return await this.retrieve_({ id: giftCardId }, config)
+    })
+  }
+
+  async retrieveByCode(
+    code: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async () => {
+      return await this.retrieve_({ code }, config)
+    })
+  }
+
+  /**
+   * Updates a giftCard.
+   * @param giftCardId - giftCard id of giftCard to update
+   * @param update - the data to update the giftCard with
+   * @return the result of the update operation
+   */
+  async update(
+    giftCardId: string,
+    update: UpdateGiftCardInput
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const giftCard = await this.retrieve(giftCardId)
+
+      const { region_id, metadata, balance, ...rest } = update
+
+      if (region_id && region_id !== giftCard.region_id) {
+        const region = await this.regionService_.retrieve(region_id)
+        giftCard.region_id = region.id
+      }
+
+      if (metadata) {
+        giftCard.metadata = await setMetadata(giftCard, metadata)
+      }
+
+      if (typeof balance !== "undefined") {
+        if (balance < 0 || giftCard.value < balance) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_ARGUMENT,
+            "new balance is invalid"
+          )
+        }
+
+        giftCard.balance = balance
+      }
+
+      for (const [key, value] of Object.entries(rest)) {
+        giftCard[key] = value
+      }
+
+      return await giftCardRepo.save(giftCard)
+    })
+  }
+
+  /**
+   * Deletes a gift card idempotently
+   * @param giftCardId - id of gift card to delete
+   * @return the result of the delete operation
+   */
+  async delete(giftCardId: string): Promise<void> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const giftCard = await giftCardRepo.findOne({ where: { id: giftCardId } })
+
+      if (!giftCard) {
+        return Promise.resolve()
+      }
+
+      await giftCardRepo.softRemove(giftCard)
+
+      return Promise.resolve()
```

Comment:
suggestion: I believe we could merge these into one
```suggestion
      return await giftCardRepo.softRemove(giftCard)
```